### PR TITLE
Fix/6

### LINKED ISSUE #6: Support for Node 10 and 12
The API for perf_hooks was changed in Node 10. We should support that as well.

### PR DESCRIPTION
**Pull request by davealbert**
Fixes #6 

Update to perf_hooks, now supporting for node 10.x and above as 8.x is no longer in LTS.

#### Checklist

- [X ] run `npm run test` and `npm run benchmark` -- Tests pass, no existing benchmarks
- [X ] tests and/or benchmarks are included -- Tests pass
- [X ] documentation is changed or added -- No existing documentation related to version support
- [X ] commit message and code follows [Code of conduct] (https://github.com/fastify/fastify/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -1,14 +1,33 @@
...
+let observedEntries = []
 
 module.exports = fp(async function (fastify, opts) {
+  const obs = new PerformanceObserver((items) => {
+    const fetchedItems = items.getEntries()
+    for (let i = 0; i < fetchedItems.length; i++) {
+      const e = fetchedItems[i]
+      if (e.name.indexOf(ROUTES) === 0) {
+        const key = e.name.split(':')[1]
+        if (observedEntries[key]) {
```
Comment (mcollina) [inline index.js]: it seems we are using `observedEntries` like an object, but it is an array instead.

Comment (davealbert) [reply index.js]: Yes, thanks, it started as an Array then moved to object, I'll fix that.

File: index.js [hunk 2]
```diff
@@ -1,14 +1,33 @@
...
 
 const fp = require('fastify-plugin')
-const { performance } = require('perf_hooks')
+const { performance, PerformanceObserver } = require('perf_hooks')
 const summary = require('summary')
 
 const ONSEND = 'on-send-'
 const PREHANDLER = 'pre-handler-'
 const ROUTES = 'fastify-routes:'
+let observedEntries = []
```
[the remark this answers is not in the document]
Comment (davealbert) [reply index.js]: As it's used in the helper functions, I'm not sure how to access it if it's in the scope of the plugin.  Am I missing something?  Should the helper functions be moved into the plugin scope?

[the remark this answers is not in the document]
Comment (mcollina) [reply index.js]: Yes exactly. They should all be moved to the plugin scope. No globals if they can be avoided.

[the remark this answers is not in the document]
Comment (davealbert) [reply index.js]: OK cool, I thought there might have been a reason for `function measurements` and `function stats` to be outside the plugin scope.

[the remark this answers is not in the document]
Comment (mcollina) [reply index.js]: In the previous implementation the performance observer was a global, so it did not matter at all. This implementation fixes that.

[the remark this answers is not in the document]
Comment (davealbert) [reply index.js]: I see, thanks for the help/feedback.

File: index.js [hunk 3]
```diff
@@ -1,14 +1,33 @@
...
+          observedEntries[key] = [e.duration]
+        }
+      }
+    }
+
+    performance.clearMarks()
+  })
+  obs.observe({ entryTypes: ['measure'] })
+
   fastify.addHook('preHandler', function (request, reply, next) {
```
Comment (mcollina) [inline index.js]: while we are updating this, we might want to move this to `onRequest`. It is more accurate.

File: index.js [hunk 4]
```diff
@@ -38,40 +57,31 @@ module.exports = fp(async function (fastify, opts) {
...
+        sd: s.sd()
+      }
 
-    return acc
-  }, {})
-}
+      return acc
+    }, {})
+  }
+})
```
Comment (mcollina) [inline index.js]: Can you set the minimum version of fastify to 2.x here? Check the docs of fastify-plugin

Comment (davealbert) [reply index.js]: This breaks a test, looking into it, but FYI may be a minor delay.

Comment (davealbert) [reply index.js]: Updated, but please check the update to the test.

Comment (mcollina) [reply index.js]: I meant something different, it should on the second argument of fp (in other terms, it should not show up in `stats()`.

Comment (davealbert) [reply index.js]: Of course, I see what I did there now.

File: index.js [hunk 5]
```diff
@@ -1,17 +1,36 @@
...
+    }
+
+    performance.clearMarks()
+  })
+  obs.observe({ entryTypes: ['measure'] })
+
+  fastify.addHook('onRequest', function (request, reply, next) {
     const id = request.raw.id
-    performance.mark(PREHANDLER + id)
+    performance.mark(ONREQUEST + id)
```
Comment (mcollina) [inline index.js]: Can you also add the request method in the key?

Comment (davealbert) [reply index.js]: If there is a `reply.context.config.statsId` eg. 'group-stats-together' would we expect this to still be separated by method, or together?  Or should we maybe add an additional config option say `reply.context.config.mergeMethods` ?  I think separated by the method makes the most sense here, but I suppose there could be a situation where one might want to join them all together.

Comment (mcollina) [reply index.js]: I think an option `mergeMethods would be good`.

**Pull request merged**